src/zarr/core/indexing.py

```diff
@@ -167,16 +167,76 @@ def _iter_grid(
         ):
             if o + ss > gs:
                 raise IndexError(
-                    f"Invalid selection shape ({selection_shape}) for origin ({origin}) and grid shape ({grid_shape}) at axis {idx}."
+                    f"Invalid selection shape ({ss}) for origin ({o}) and grid shape ({gs}) at axis {idx}."
                 )
             dimensions += (range(o, o + ss),)
-        yield from itertools.product(*(dimensions))
+        return itertools.product(*(dimensions))
 
     else:
         msg = f"Indexing order {order} is not supported at this time."  # type: ignore[unreachable]
         raise NotImplementedError(msg)
 
 
+def iter_regions(
```

Comment:
this is a new function for iterating over contiguous regions. When we support irregular chunking, we can overload the type of `region_shape` accordingly.